### PR TITLE
core: add --no-fluffy-blocks, and enable fluffy blocks by default

### PR DESCRIPTION
**Pull request by moneromooo-monero**

### CODE REVIEW

**Comment by NanoAkron** [comment]
@jtgrassie I think it's worth the upgrade. If people can't be bothered to do a one-time change to their init scripts...hmm...

**Comment by jtgrassie** [comment]
@NanoAkron I'm not against the upgrade, just don't see the point breaking existing functionality. The upgrade (to default fluffy blocks) can occur without removing the existing option.

**Comment by stoffu** [comment]
@jtgrassie It's not breaking any existing functionality. The flag `--no-fluffy-blocks` alone suffices, and the existing flag `--fluffy-blocks` becomes redundant. I don't think it's a good idea to bloat the set of flags (that is already quite big) just for the sake of keeping users' init scripts backward compatible.

**Comment by jtgrassie** [comment]
@stoffu fair point. If it doesn't _break_ running with --fluffy-blocks then no problem.

**Comment by stoffu** [comment]
@jtgrassie Well, the idea here is to *remove* `--fluffy-blocks`, so it'll be considered as unrecognized flag and the program halts with the help message. Existing init scripts do need to delete `--fluffy-blocks`. Do you think this is problematic?

**Comment by hyc** [comment]
Would be better IMO to just keep the --fluffy-blocks parameter, but make it take an optional on/off argument. (Defaulting on, if not specified.) It was originally supposed to behave that way anyway.

**Comment by jtgrassie** [comment]
@stoffu that's what I'm worried about. I'd prefer a depreciation message than a halt.

**Comment by jtgrassie** [comment]
i feel @hyc has the most appropriate suggestion. This won't break existing invocations and change the default. Winner winner.

**Comment by stoffu** [comment]
I also like @hyc's suggestion.

**Comment by hyc** [comment]
Something like this instead:
````
diff --git a/src/cryptonote_core/cryptonote_core.cpp b/src/cryptonote_core/cryptonote_core.cpp
index adbc727..c5b13c3 100644
--- a/src/cryptonote_core/cryptonote_core.cpp
+++ b/src/cryptonote_core/cryptonote_core.cpp
@@ -128,10 +128,10 @@ namespace cryptonote
   , "Check for new versions of monero: [disabled|notify|download|update]"
   , "notify"
   };
-  static const command_line::arg_descriptor<bool> arg_fluffy_blocks  = {
+  static const command_line::arg_descriptor<uint64_t> arg_fluffy_blocks  = {
     "fluffy-blocks"
-  , "Relay blocks as fluffy blocks where possible (automatic on testnet)"
-  , false
+  , "Relay blocks as fluffy blocks where possible"
+  , 1
   };
 
   //-----------------------------------------------------------------------------------------------
@@ -268,7 +268,7 @@ namespace cryptonote
 
     set_enforce_dns_checkpoints(command_line::get_arg(vm, arg_dns_checkpoints));
     test_drop_download_height(command_line::get_arg(vm, arg_test_drop_download_height));
-    m_fluffy_blocks_enabled = m_testnet || get_arg(vm, arg_fluffy_blocks);
+    m_fluffy_blocks_enabled = get_arg(vm, arg_fluffy_blocks) != 0;
     m_offline = get_arg(vm, arg_offline);
 
     if (command_line::get_arg(vm, arg_test_drop_download) == true)
````

**Comment by moneromooo-monero** [comment]
The above will break scripts anyway AFAICT since it requires --fluffy-blocks 1 in order to work, and any such script has just --fliuffy-blocks.

**Comment by stoffu** [comment]
Ah, a flag with `arg_descriptor<uint64_t>` can't omit the actual integral value after the flag name? Bummer :(

**Comment by jtgrassie** [comment]
You sure? `arg_descriptor` has a `not_use_default`, and `get_arg` has a `required` flag. Some combination of these maybe?

**Comment by stoffu** [comment]
@jtgrassie I applied @hyc's patch and confirmed what @moneromooo-monero stated. Tweaking `not_use_default` didn't help either.

**Comment by jtgrassie** [comment]
Given the current `arg_descriptor<bool> arg_fluffy_blocks` is a bool, why can't that just be left alone? Then if the arg is not present, default to `true`.

**Comment by jtgrassie** [comment]
Something like:
```
diff --git a/src/cryptonote_core/cryptonote_core.cpp b/src/cryptonote_core/cryptonote_core.cpp
index 81a7b472..c1896296 100644
--- a/src/cryptonote_core/cryptonote_core.cpp
+++ b/src/cryptonote_core/cryptonote_core.cpp
@@ -135,7 +135,7 @@ namespace cryptonote
   static const command_line::arg_descriptor<bool> arg_fluffy_blocks  = {
     "fluffy-blocks"
   , "Relay blocks as fluffy blocks where possible (automatic on testnet)"
-  , false
+  , true
   };
 
   //-----------------------------------------------------------------------------------------------
@@ -273,7 +273,7 @@ namespace cryptonote
 
     set_enforce_dns_checkpoints(command_line::get_arg(vm, arg_dns_checkpoints));
     test_drop_download_height(command_line::get_arg(vm, arg_test_drop_download_height));
-    m_fluffy_blocks_enabled = m_testnet || get_arg(vm, arg_fluffy_blocks);
+    m_fluffy_blocks_enabled = get_arg(vm, arg_fluffy_blocks);
     m_offline = get_arg(vm, arg_offline);
     m_disable_dns_checkpoints = get_arg(vm, arg_disable_dns_checkpoints);
```

**Comment by stoffu** [comment]
@jtgrassie I've just tried that now, and it doesn't work.

**Comment by jtgrassie** [comment]
Odd the default is not being picked up. How about this:

```
diff --git a/src/cryptonote_core/cryptonote_core.cpp b/src/cryptonote_core/cryptonote_core.cpp
index 81a7b472..d8706f1d 100644
--- a/src/cryptonote_core/cryptonote_core.cpp
+++ b/src/cryptonote_core/cryptonote_core.cpp
@@ -135,7 +135,7 @@ namespace cryptonote
   static const command_line::arg_descriptor<bool> arg_fluffy_blocks  = {
     "fluffy-blocks"
   , "Relay blocks as fluffy blocks where possible (automatic on testnet)"
-  , false
+  , true
   };
 
   //-----------------------------------------------------------------------------------------------
@@ -273,7 +273,9 @@ namespace cryptonote
 
     set_enforce_dns_checkpoints(command_line::get_arg(vm, arg_dns_checkpoints));
     test_drop_download_height(command_line::get_arg(vm, arg_test_drop_download_height));
-    m_fluffy_blocks_enabled = m_testnet || get_arg(vm, arg_fluffy_blocks);
+    m_fluffy_blocks_enabled = true;
+    if (command_line::has_arg(vm, arg_fluffy_blocks))
+        m_fluffy_blocks_enabled = command_line::get_arg(vm, arg_fluffy_blocks);
     m_offline = get_arg(vm, arg_offline);
     m_disable_dns_checkpoints = get_arg(vm, arg_disable_dns_checkpoints);
```

**Comment by stoffu** [comment]
@jtgrassie Did you try it yourself? As far as I tested it, it's now always enabled and impossible to disable it.

**Comment by jtgrassie** [comment]
`--fluffy-blocks=false` to disable? And sorry no, not tested as I'm on the move.

**Comment by jtgrassie** [comment]
@moneromooo-monero makes sense.

**Comment by NanoAkron** [comment]
This is silly. All 6 people who don't want to update their init scripts will eventually figure it out.

**Comment by jtgrassie** [comment]
@NanoAkron which 6 people? You know all the people who this might affect? Doesn't affect me but I assume it will affect others. IMHO a nice depreciation path is preferable to breaking changes. But I do agree it's a little silly as it doesn't have huge implications ;)

File: src/cryptonote_core/cryptonote_core.cpp [hunk 1]
```diff
@@ -268,8 +274,10 @@ namespace cryptonote
 
     set_enforce_dns_checkpoints(command_line::get_arg(vm, arg_dns_checkpoints));
     test_drop_download_height(command_line::get_arg(vm, arg_test_drop_download_height));
-    m_fluffy_blocks_enabled = m_testnet || get_arg(vm, arg_fluffy_blocks);
+    m_fluffy_blocks_enabled = !get_arg(vm, arg_no_fluffy_blocks);
     m_offline = get_arg(vm, arg_offline);
+    if (!command_line::is_arg_defaulted(vm, arg_fluffy_blocks))
+      MWARNING(arg_fluffy_blocks.name << " is obsolete, it is now default");
```
Comment (NanoAkron) [inline src/cryptonote_core/cryptonote_core.cpp]: Worth adding something like 'remove this line from your init script'?

**Pull request merged**